app/routes.py

```diff
@@ -1,2 +1,270 @@
+    elif request.method == "PUT":
+    # form data is a local variable to hold the body of the HTTP request
+        form_data = request.get_json()
+    # reassigning attributes of the task object using the dict values that came over 
+    # in the request body
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = task.completed_at
+
+        db.session.commit()
+
+        return make_response({"task": {"id": task.task_id,
+                                    "title": task.title,
+                                    "description": task.description,
+                                    "is_complete": is_complete}}, 200)
+    # patch will change just one part of the record, not the whole record
+    elif request.method == "PATCH":
+        form_data = request.get_json()
+        if "title" in form_data:
+            task.title = form_data["title"]
+        if "description" in form_data:
+            task.description = form_data["description"]
+        
+        db.session.commit()
+        return make_response({"task": {"id": task.task_id,
+                                    "title": task.title,
+                                    "description": task.description,
+                                    "is_complete": is_complete}}, 200)
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return make_response({'details': 
+        f'Task {task.task_id} "{task.title}" successfully deleted'}, 200)
+        
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_task_complete(task_id):
+# setting current datetime to time_stamp to use to mark task.completed_at
+    time_stamp = datetime.now()
+    task = Task.query.get(task_id) #either get Task back or None
+    if task is None:
+        abort(404) 
+    else:
+        task.completed_at = time_stamp
+# now db record shows timestamp for completed_at
+    db.session.commit()
+    is_complete = task.completed_at is not None
+    
+    try:
+        query_params = {"channel": "slack-api-test-channel",
+        "text": f"Someone just completed the task {task.title}"}
+        header_param = {"Authorization": "Bearer "+ os.environ.get("slack_oauth_token")}
+        slack_post_body = requests.post(slack_path, params=query_params, headers= header_param)
+    except TypeError:
+        pass
+# response body doesn't want to show timestamp, just True, so changing is_complete to True
+    return make_response({"task": {"id": task.task_id,
+                                    "title": task.title,
+                                    "description": task.description,
+                                    "is_complete": is_complete}}, 200)
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete(task_id):
+    task = Task.query.get(task_id) #either get Task back or None
+    if task is None:
+        abort(404) 
```

Comment:
another suggestion instead of abort could be 
```suggestion
    if not task:
        return "", 404
```